core/trino-main/src/main/java/io/trino/operator/output/RowPositionsAppender.java

```diff
@@ -0,0 +1,235 @@
+import static io.trino.operator.output.PositionsAppenderUtil.calculateNewArraySize;
+import static io.trino.spi.block.RowBlock.createRowBlockInternal;
+import static java.util.Objects.requireNonNull;
+
+public class RowPositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(RowPositionsAppender.class).instanceSize();
+    private final PositionsAppender[] fields;
+    private int initialEntryCount;
+    private boolean initialized;
+
+    private int positionCount;
+    private boolean hasNullRow;
+    private boolean hasNonNullRow;
+    private boolean[] rowIsNull = new boolean[0];
+    private int[] fieldBlockOffsets = new int[1];
+    private long retainedSizeInBytes;
+    private long sizeInBytes;
+
+    public static RowPositionsAppender createRowAppender(
+            PositionsAppenderFactory positionsAppenderFactory,
+            RowType type,
+            int expectedPositions,
+            long maxPageSizeInBytes)
+    {
+        PositionsAppender[] fields = new PositionsAppender[type.getFields().size()];
+        for (int i = 0; i < fields.length; i++) {
+            fields[i] = positionsAppenderFactory.create(type.getFields().get(i).getType(), expectedPositions, maxPageSizeInBytes);
+        }
+        return new RowPositionsAppender(fields, expectedPositions);
+    }
+
+    private RowPositionsAppender(PositionsAppender[] fields, int expectedPositions)
+    {
+        this.fields = requireNonNull(fields, "fields is null");
+        this.initialEntryCount = expectedPositions;
+        updateRetainedSize();
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block block)
+    {
+        if (positions.isEmpty()) {
+            return;
+        }
+        ensureCapacity(positions.size());
+        RowBlock sourceRowBlock = (RowBlock) block;
+        IntArrayList notNullPositions;
+        if (sourceRowBlock.mayHaveNull()) {
+            notNullPositions = processNullablePositions(positions, sourceRowBlock);
+            hasNullRow |= notNullPositions.size() < positions.size();
+            hasNonNullRow |= notNullPositions.size() > 0;
+        }
+        else {
+            // the source Block does not have nulls
+            notNullPositions = positions;
+
+            for (int i = 0; i < positions.size(); i++) {
+                fieldBlockOffsets[positionCount + i + 1] = fieldBlockOffsets[positionCount] + i + 1;
+            }
+            hasNonNullRow = true;
+        }
+
+        sourceRowBlock.writeRowsTo(
+                notNullPositions.elements(),
+                0,
+                notNullPositions.size(),
+                (fieldBlocks, mappedFieldPositions, positionCount) -> {
+                    IntArrayList mappedPositions = IntArrayList.wrap(mappedFieldPositions, positionCount);
+                    for (int i = 0; i < fields.length; i++) {
+                        fields[i].append(mappedPositions, fieldBlocks[i]);
```

Comment:
Is it possible that the `fieldBlocks` coming out of RowBlock is a LazyBlock ? Not sure if PositionsAppender deals with those